src/vector/rageshake.js

```diff
@@ -0,0 +1,409 @@
+            req.onsuccess = (event) => {
+                this.db = event.target.result;
+                // Periodically flush logs to local storage / indexeddb
+                setInterval(this.flush.bind(this), FLUSH_RATE_MS);
+                resolve();
+            };
+
+            req.onerror = (event) => {
+                const err = "Failed to open log database: " + event.target.errorCode;
+                console.error(err);
+                reject(new Error(err));
+            };
+
+            // First time: Setup the object store
+            req.onupgradeneeded = (event) => {
+                const db = event.target.result;
+                const objectStore = db.createObjectStore("logs", {
+                    keyPath: ["id", "index"]
+                });
+                // Keys in the database look like: [ "instance-148938490", 0 ]
+                // Later on we need to query for everything with index=0, and query everything for an instance id.
+                // In order to do this, we need to set up indexes on both "id" and "index".
+                objectStore.createIndex("index", "index", { unique: false });
+                objectStore.createIndex("id", "id", { unique: false });
+
+                objectStore.add(
+                    this._generateLogEntry(
+                        new Date() + " ::: Log database was created."
+                    )
+                );
+            }
+        });
+    }
+
+    /**
+     * @return {Promise} Resolved when the logs have been flushed.
+     */
+    flush() {
+        if (!this.db) {
+            // not connected yet or user rejected access for us to r/w to the db
+            return Promise.reject(new Error("No connected database"));
+        }
+        const lines = this.logger.flush();
+        if (lines.length === 0) {
+            return Promise.resolve();
+        }
+        return new Promise((resolve, reject) => {
+            let txn = this.db.transaction("logs", "readwrite");
+            let objStore = txn.objectStore("logs");
+            objStore.add(this._generateLogEntry(lines));
+            txn.oncomplete = (event) => {
+                resolve();
+            };
+            txn.onerror = (event) => {
+                console.error("Failed to flush logs : " + event.target.errorCode);
+                reject(new Error("Failed to write logs: " + event.target.errorCode));
+            }
+        });
+    }
+
+    /**
+     * Consume the most recent logs and return them. Older logs which are not returned are deleted at the same time,
+     * so this can be called at startup to do house-keeping to keep the logs from growing too large.
+     *
+     * @param {boolean} clearAll True to clear the most recent logs returned in addition to the
+     * older logs. This is desirable when sending bug reports as we do not want to submit the
+     * same logs twice. This is not desirable when doing house-keeping at startup in case they
```

Comment:
I don't follow. At startup, all we want to do is clear old logs, not **all** logs since we're just doing housekeeping. When rageshakes are submitted, we want to clear **all** logs because otherwise you'll get duplicates if they submit a 2nd or 3rd bug report.